core/schema/module.go

```diff
@@ -316,176 +333,841 @@ func (s *moduleSchema) functionWithArg(ctx context.Context, fn *core.Function, a
-			}
-			configPath, err = filepath.Rel(rootPathAbs, configPathAbs)
-			if err != nil {
-				return nil, fmt.Errorf("failed to get config relative to root: %w", err)
-			}
-			if strings.HasPrefix(configPath, "../") {
-				// this likely shouldn't happen, a client shouldn't submit a
-				// module config that escapes the module root
-				return nil, fmt.Errorf("module subpath is not under module root")
-			}
-			if rootPath != "." {
-				err = s.dag.Select(ctx, sourceDir, &sourceDir, dagql.Selector{
-					Field: "directory",
-					Args: []dagql.NamedInput{
-						{Name: "path", Value: dagql.String(rootPath)},
-					},
-				})
-				if err != nil {
-					return nil, fmt.Errorf("failed to get root directory: %w", err)
-				}
-			}
+func (s *moduleSchema) moduleWithSource(ctx context.Context, self dagql.Instance[*core.Module], args struct {
+	Source core.ModuleSourceID
+}) (*core.Module, error) {
+	src, err := args.Source.Load(ctx, s.dag)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode module source: %w", err)
+	}
+
+	src, rootDir, sourceSubpath, err := s.normalizeSourceForModule(ctx, src)
+	if err != nil {
+		return nil, fmt.Errorf("failed to normalize module source: %w", err)
+	}
+
+	var modsCfg modules.ModulesConfig
+	configFile, err := rootDir.Self.File(ctx, filepath.Join("/", modules.Filename))
+	if err == nil {
+		configBytes, err := configFile.Contents(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read module config file: %w", err)
+		}
+
+		if err := json.Unmarshal(configBytes, &modsCfg); err != nil {
+			return nil, fmt.Errorf("failed to decode module config: %w", err)
+		}
+		if err := modsCfg.Validate(); err != nil {
+			return nil, fmt.Errorf("invalid module config: %w", err)
+		}
+	}
+	relSubpath, err := filepath.Rel("/", sourceSubpath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get module source path relative to config: %w", err)
+	}
+
+	modCfg, foundModCfg := modsCfg.ModuleConfigByPath(relSubpath)
+	if !foundModCfg {
+		// no config for this module found, need to initialize a new one
+		modCfg = &modules.ModuleConfig{
+			Source: relSubpath,
 		}
 	}
 
-	sourceDirSubpath := filepath.Dir(configPath)
+	mod := self
+	mod.Self = mod.Self.Clone()
+	mod.Self.Source = src
+	mod.Self.Source.Self = mod.Self.Source.Self.Clone()
+	mod.Self.NameField = modCfg.Name
+	mod.Self.SDKConfig = modCfg.SDK
+	mod.Self.GeneratedSourceDirectory = rootDir
+	mod.Self.GeneratedSourceDirectory.Self = rootDir.Self.Clone()
+	mod.Self.GeneratedSourceSubpath = sourceSubpath
+	mod.Self.DirectoryIncludeConfig = modCfg.Include
+	mod.Self.DirectoryExcludeConfig = modCfg.Exclude
+
+	// honor the include/exclude values for the directory (even if it's too late to actually use them
+	// for loading performance optimizations)
+	if len(modCfg.Include) > 0 || len(modCfg.Exclude) > 0 {
+		mod.Self.GeneratedSourceDirectory.Self, err = mod.Self.GeneratedSourceDirectory.Self.WithDirectory(
```

Comment:
> If I'm reading this right, it's modifying a dagql.Instance but leaving its ID as it originally was, which seems subtly broken. Similar weird feeling about updating mod.Self but it doesn't matter because we just return mod.Self in the end anyhow.

Ah right, good catch! I think I can update this to use `s.dag.Select` instead.

> Maybe dagql.Instance should make Self a read-only getter? 🤔 🤔

I think something like that could help a lot, though I suppose we'd need to make sure that when you do `dag.Select(ctx, foo, &theDest)` then `theDest` still gets the updated `Self` (whatever field that's stored in).

Throughout this all I also kept *feeling* like I wanted to just modify `Self` and get the updated ID reflected in the instance automatically, but that being said:
1. I don't know how to implement in a clean way
2. Recalculating IDs super often might be an unobvious performance hazard
3. Dealing with the fact that you need to use `Select` to get updated `Instance`s seemed to result in better overall structure to the API once all was said and done, due to the fact that it required there actually being an API to `Select`.
   * It admittedly was a bit annoying at times since it added time+mental overhead to every iteration of modifications I made to the implementation here, but I think that was exacerbated by the fact that these API updates in particular just required 99999+ iterations to get right, not sure if it's really going to be a problem very often or if this was just an exceptional case.

Reply:
Fixed